### DISCUSSION TITLE
How to listen to the right click event on the canvas?

### DISCUSSION THREAD

**Discussion by kamalesh-d** [GrapesJS/grapesjs#4844] [General]
Hello, 

First of all thank you very much for making such great project in open source. 

Can someone please help me, what is the way to listen to the right-click event on grapesjs canvas? I wanted to listen to this event and open a context-menu for designing the element at which the click was performed. 

I learnt to get the current selected component from the canvas based on the following code: 
 
editor.on('component:selected', (selectedcomponent) => {
   \# Selected component can be processed as I want.
});

However, I am not able to listen to the right-click event. I am working with Vuejs, what I tried is:

## HTML content
        <div id="gjs" v-html="current_page.content" @contextmenu.prevent="rightClickHandle"></div>

## Vue method
rightClickHandle(){
  alert('right clicked');      
}

Normally, it shall be able to detect the right click in this way. But since the canvas is attached through 'gjs' id, I need to know if there is a way grapesjs (API) allows us to listen for the right-click event. 

Please let me know if any additional information is required. 

Thanks in advance, 
Kamalesh D.

**Comment by artf** [comment]
You can get the iframe document from the canvas and catch your events
```js
editor.Canvas.getDocument().addEventListener('contextmenu', (ev) => {
    ev.preventDefault();
    console.log('contextmenu inside canvas iframe');
})
```

**Reply by kamalesh-d** [reply]
Hello @artf,

Thank you very much for your quick response. 

I tried the solution, but I didn't get successful in it. I am sorry if I am making any silly mistakes on it.

Please find the JSfiddle I have prepared with your solution as below:
https://jsfiddle.net/hvfs4aed/9/

When I try to addEventListener for the whole window, it works fine (which is commented in the fiddle). When I test the same through Canvas.getDocument() as per your suggestion, it doesn't work. 

Please guide me on this, if I need to make any modifications. 

Thanks again for your help!!!

**Reply by artf** [reply]
In your demo update the height of the editor (eg. `height: '500px',`) in order to see the example working.

**Reply by kamalesh-d** [reply]
Hello @artf,

I tried your solution, however, it didn't help me solve it. 

https://jsfiddle.net/fs6bdwvh/2/

Could there any problem of z-index, by which canvas elements are behind some other div? 

And when I search more, I found similar solution proposed for the double-click listener in StackOverflow. However, I am not able to listen to any event. 
https://stackoverflow.com/questions/68175490/trap-on-dbclick-in-grapejs-editor

Thanks, 
Kamalesh D.

**Reply by artf** [reply]
I see it working but I guess there might be some timing issue as iframes are actually rendered asynchronously. Try to attach the listener once the editor is ready
```js
editor.onReady(() => {
  editor.Canvas.getDocument().addEventListener('contextmenu', (ev) => {...});
})
```

**Reply by kamalesh-d** [reply]
That resolved my issue, thank you very much for your help. Really very kind. 

Regards, 
Kamalesh D.

**Reply by usmanqamar-int** [reply]
> ```js
> editor.onReady(() => {
>   editor.Canvas.getDocument().addEventListener('contextmenu', (ev) => {...});
> })
> ```

Thanks @artf . It's working for me.

**Reply by badiuciprian** [reply]
Cool, it works but is there an eventListener for **some** components only? For example I need to add different actions for different components, and not have a context on the main wrapper

**Reply by ronaldohoch** [reply]
@badiuciprian i think you can use the `editor.getSelected()` and check the informations to show custon buttons of the context menu.

But o don't know if right click selects the components in Canvas 🤔

**Reply by badiuciprian** [reply]
as I tried so far, the context opens if you click anywhere on the canvas, because of the wrapper container which is basically a component and I could not make it "ignorable"
